### PR TITLE
Publish DNS records for ingress resources in KubeDNS.

### PR DESCRIPTION
**Pull request by madhusudancs**
Fixes issue #29345

E2E tests on the way.

``` release-note
KubeDNS now publishes DNS A/CNAME records for Ingress resources and can thus be used for Ingress discovery. The domain names are of the form `<ingressname>.<namespace>.ing.cluster.local.`. Clients can drop the `.cluster.local.` part in their DNS requests. 

Since Ingress only supports HTTP and HTTPS through ports 80 and 443 respectively for now, we do not publish SRV records for these resources. Please let us know if this is important for you.
```

cc @kubernetes/sig-cluster-federation

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/29768)

### CODE REVIEW

File: pkg/dns/dns.go [hunk 1]
```diff
@@ -268,6 +301,97 @@ func (kd *KubeDNS) updateService(oldObj, newObj interface{}) {
...
+		subCache := NewTreeCache()
+		subCachePath := append(kd.domainPath, ingressSubdomain, ing.Namespace)
+
+		ips := []string{}
+		ingFQDN := strings.Join([]string{ing.Name, ing.Namespace, ingressSubdomain, kd.domain}, ".")
+		reverseRecord, _ := getSkyMsg(ingFQDN, 0)
+		for _, lbi := range ing.Status.LoadBalancer.Ingress {
+			// We generate an A record if the Ingress loadbalancer IP is set and a
+			// CNAME record if the hostname is set. If both IP and hostname fields
+			// are set we favor IP over the hostname.
```
Comment (nikhiljindal) [inline pkg/dns/dns.go]: Is there a reason for this?

File: pkg/dns/dns.go [hunk 2]
```diff
@@ -268,6 +301,97 @@ func (kd *KubeDNS) updateService(oldObj, newObj interface{}) {
...
+			// TODO(madhusudancs): We would like to generate an empty record
```
Comment (nikhiljindal) [inline pkg/dns/dns.go]: Do you want to file an issue to track this. upto you

File: pkg/dns/dns.go [hunk 3]
```diff
@@ -268,6 +301,97 @@ func (kd *KubeDNS) updateService(oldObj, newObj interface{}) {
...
+			// (NODATA response) if neither of these fields are set to indicate
+			// that we know that the Ingress resource exists but we do not know
+			// what it points to. It is not entirely clear how to make that work
+			// with SkyDNS because setting an empty host field in the service
+			// record seems to return all the A records as response. Need to
+			// figure this out.
+			if len(lbi.IP) > 0 {
+				kd.newIngressRecord(subCache, ing.Namespace, ing.Name, lbi.IP)
```
Comment (nikhiljindal) [inline pkg/dns/dns.go]: how about adding a log here stating that we are generating A record

File: pkg/dns/dns.go [hunk 4]
```diff
@@ -268,6 +301,97 @@ func (kd *KubeDNS) updateService(oldObj, newObj interface{}) {
...
+				ips = append(ips, lbi.IP)
+			} else if len(lbi.Hostname) > 0 {
+				kd.newIngressRecord(subCache, ing.Namespace, ing.Name, lbi.Hostname)
```
Comment (nikhiljindal) [inline pkg/dns/dns.go]: same, add a log

File: pkg/dns/dns.go [hunk 5]
```diff
@@ -268,6 +301,97 @@ func (kd *KubeDNS) updateService(oldObj, newObj interface{}) {
...
+			} else {
+				glog.V(4).Infof("Ingress resource with neither an IP address nor a hostname: %+v", ing)
```
Comment (nikhiljindal) [inline pkg/dns/dns.go]: How about increasing the log severity?
This loop should not be processed at all when the resource is just created, status will be empty in that case, right?

File: pkg/dns/dns.go [hunk 6]
```diff
@@ -268,6 +301,97 @@ func (kd *KubeDNS) updateService(oldObj, newObj interface{}) {
...
+		defer kd.cacheLock.Unlock()
+		kd.cache.deletePath(subCachePath...)
+		for _, ip := range ips {
+			delete(kd.reverseRecordMap, ip)
+		}
+	}
+}
+
+func (kd *KubeDNS) updateIngress(oldObj, newObj interface{}) {
+	// TODO(madhusudancs): Ensure we don't leave the records dangling from the
```
Comment (nikhiljindal) [inline pkg/dns/dns.go]: How about calling removeIngress(oldObj) and then newIngress(newObj)

**Comment by nikhiljindal** [comment]
Would be good if @bprashanth or @girishkalele takes a look once as well, since they know KubeDNS code best.

File: pkg/dns/dns.go (hunk 1 above)
Comment (madhusudancs) [inline pkg/dns/dns.go]: DNS spec doesn't allow mixing up A and CNAME records for a given domain name. So we have to pick one or the other. And IP address is more specific than a CNAME because CNAME should ultimately resolve to some IP address, so choosing this order.

Comment (madhusudancs) [inline pkg/dns/dns.go]: Also, added this comment in the code. Thanks for pointing it out.

File: pkg/dns/dns.go (hunk 2 above)
Comment (madhusudancs) [inline pkg/dns/dns.go]: Not entirely sure if this is worth an issue. Leaving it here for now.

File: pkg/dns/dns.go (hunk 3 above)
Comment (madhusudancs) [inline pkg/dns/dns.go]: `kd.newIngressRecord()` calls `getSkyMsg()` which generates a log message for this. And it will be evident in that log that this will be an A record.

File: pkg/dns/dns.go (hunk 4 above)
Comment (madhusudancs) [inline pkg/dns/dns.go]: Same as above.

File: pkg/dns/dns.go (hunk 1 above)
Comment (madhusudancs) [inline pkg/dns/dns.go]: Actually thank you very much for asking for the reason here. This got me thinking and made me find a potential bug here which might have happened in some very rare corner cases. Refactored the code a bit and added more tests.

File: pkg/dns/dns.go (hunk 5 above)
Comment (madhusudancs) [inline pkg/dns/dns.go]: Done. Promoted it to a `Warning`.

File: pkg/dns/dns.go (hunk 6 above)
Comment (madhusudancs) [inline pkg/dns/dns.go]: It is not necessary, as in unnecessary computation. `newIngress()` overwrites the corresponding tree node. That TODO was more of a self-note to verify that this was actually true. I have removed the TODO now. Thanks pointing it out though.

**Comment by madhusudancs** [comment]
@nikhiljindal Addressed the comments, PTAL.

@girishkalele could you please review this PR?

**Comment by girishkalele** [comment]
@madhusudancs I will take a look today.

File: pkg/dns/dns.go [hunk 7]
```diff
@@ -268,6 +301,124 @@ func (kd *KubeDNS) updateService(oldObj, newObj interface{}) {
...
+// If a CNAME RR is present at a node, no other data should be
+// present; this ensures that the data for a canonical name and its
+// aliases cannot be different.
+// """
+// Due to this, we can either only generate A records or a CNAME
+// record but not both. We have to choose between the two. IP addresses
+// are more specific than a CNAME because a CNAME should eventually
+// resolve to an IP address, so favoring IP addresses over CNAME.
+//
+// TODO(madhusudancs): We would like to generate an empty record
```
Comment (girishkalele) [inline pkg/dns/dns.go]: TODO ?

File: pkg/dns/dns.go [hunk 8]
```diff
@@ -268,6 +301,124 @@ func (kd *KubeDNS) updateService(oldObj, newObj interface{}) {
...
+// (NODATA response) if neither of these fields are set to indicate
+// that we know that the Ingress resource exists but we do not know
+// what it points to. It is not entirely clear how to make that work
+// with SkyDNS because setting an empty host field in the service
+// record seems to return all the A records as response. Need to
+// figure this out.
+func (kd *KubeDNS) newIngress(obj interface{}) {
```
Comment (girishkalele) [inline pkg/dns/dns.go]: What are we returning today in this case ? NXDOMAIN ?

Comment (madhusudancs) [inline pkg/dns/dns.go]: Yeah NXDOMAIN.

File: pkg/dns/dns.go (hunk 7 above)
Comment (madhusudancs) [inline pkg/dns/dns.go]: @girishkalele did not understand this question. Could you please elaborate?

File: pkg/dns/dns.go [hunk 9]
```diff
@@ -268,6 +301,124 @@ func (kd *KubeDNS) updateService(oldObj, newObj interface{}) {
...
+
+// newIngress generates DNS records for an Ingress resource.
+//
+// If all the Ingress loadbalancer status entries have their IP fields
+// set we generate a series of A records for them, one per IP. If there
+// is only one hostname in the Ingress status entries, we generate a
+// CNAME record for it. If there are multiple hostnames and no IP then
+// we generate a CNAME for the first hostname. If both IPs and hostnames
+// coexist, we favor IPs over the hostnames and we generate only A
+// records for the IPs and no CNAME record at all.
```
Comment (nikhiljindal) [inline pkg/dns/dns.go]: For my own curiosity, whats a valid use case when an ingress will have multiple LoadbalancerIngresses with different IPs and hostnames?

Comment (madhusudancs) [inline pkg/dns/dns.go]: An Ingress implementation could provide both a set of IP address endpoints and automatically create a DNS entry for it and put them down in the status fields. I don't think we have any implementation today that use these fields in this way, but it is entirely possible.

File: pkg/dns/dns.go [hunk 10]
```diff
@@ -51,6 +52,9 @@ const (
 	// A subdomain added to the user specified dmoain for all pods.
 	podSubdomain = "pod"
 
+	// A subdomain added to the user specified domain for all ingresses.
+	ingressSubdomain = "ing"
```
Comment (quinton-hoole) [inline pkg/dns/dns.go]: @thockin @bprashanth Any objections to this?  It's the equivalent to '.svc' for services. What process do we need to follow to take this out of Alpha?  Is it a big enough decision to drag through the networking/API SIG's?  I suspect not.

**Comment by nikhiljindal** [comment]
Thanks for the clarification @madhusudancs .
LGTM.

Please wait for LGTM from @girishkalele as well.

File: pkg/dns/dns.go (hunk 7 above)
Comment (girishkalele) [inline pkg/dns/dns.go]: I meant, is there an issue to track this future TODO item ?

Comment (madhusudancs) [inline pkg/dns/dns.go]: @girishkalele There isn't. @nikhiljindal also pointed that out. I don't think it is that important enough to be tracked in an issue. But if you both insist, I will open an issue.

**Comment by girishkalele** [comment]
Waiting for @thockin to OK the new .ing DNS subdomain.

**Comment by thockin** [comment]
Is thsi urgent for 1.4?  I have higher priorities and there's an
unfortunate couple of concerns I have with this, but I have not given it
deep review yet.

**Comment by quinton-hoole** [comment]
@thockin Federated Ingress is a blocker for v1.4, and this is pretty much the only think outstanding for that.  So yes, in that sense it's urgent for v1.4.  On the basis of our brief conversation last week, I can give it some further thought and put together a few concrete proposals for us to choose between then week, if that would help.

**Comment by smarterclayton** [comment]
@kubernetes/rh-networking FYI on this

**Comment by quinton-hoole** [comment]
I believe that @thockin's concerns have been/are being addressed in #30544.  @thockin please shout if you have concerns over and above what's being dealt with in #30544 (The "ing" thread).  In them mean time I suggest getting this merged and tested.  @girishkalele @thockin will one of you please apply LGTM of you're OK with that?

**Comment by quinton-hoole** [comment]
Awaiting LGTM from @thockin or @girishkalele

**Comment by quinton-hoole** [comment]
Ok, so it turns out that there is another, fully legitimate concern regarding this, which I've discussed with @thockin and @bprashanth.  I will do my best to summarize the outcome of that discussion here.  See https://github.com/kubernetes/kubernetes/issues/30544#issuecomment-241066735 for some additional context.

1. Ingress rule config has an optional 'host' construct (e.g. foo.bar.com).  Today, to make that useful for browser clients, it's necessary to manually (or at least external to Kubernetes) configure an external DNS server with an A record for that hostname pointing to the external IP of the the Ingress, otherwise the HTTP request arriving at the Ingress won't have the required hostname in it's HTTP header, and won't match the rule.
2. For similar reasons, if we automatically configure and advertise DNS names for ingress (e.g. myingress.mynamespace.ing.myfederation.foobar.com) it is not useful/usable unless either:
    1. That hostname matches a hostname in the ingress rule config
    2. The ingress rule config omits hostname and thus matches all incoming hostname headers.
    3. The user manually configures an external DNS server with a CNAME record for the hostnames configured in the ingress rule config, pointing to the auto-configured "ing" DNS record for the ingress.
    4. The user creates an A record as per 1 above (status quo).

So one counter-argument for auto-configuring and advertising these "ing" DNS records at all might be that they don't buy us much if anything.  However, on closer inspection, there seem to be some real benefits (using the numbering under 2 above):

1. Configuring the host construct in the ingress config is relatively easy (the name of the ingress and the host construct reside in the same ingress spec/yaml file.  So making them consistent is fairly natural).  And that makes accessing the ingress via a DNS name trivially easy (which is strictly better than requiring access via the IP only)
2. Ingress rules which omit hostnames are also automatically accessible via the auto-configured "ing" DNS name (strictly better than IP only)
3. CNAME records are easier to configure than A records, as the destination DNS name of the CNAME record can always be known in advance (it's derived directly from the ingress name, namespace and a configured DNS zone), and is portable across cloud providers, whereas the IP address is not always known in advance, and not portable across cloud providers.
4. No change here - status quo.

In addition (and this is somewhat orthogonal), in future we could automatically configure and manage DNS names to correctly match all hostname constructs in the Ingress configurations.  This would be a Beautiful Thing.  Right now we already have DNS provider libraries for Google Cloud DNS, AWS Route53 and SkyDNS.  Adding additional libraries is a minor job (can be done in a day or two).

So in summary, I think that there are clear benefits to adding these "ing" DNS names for Ingress.

Q

**Comment by madhusudancs** [comment]
Superseded by PR #31040.
